Progress Report - November 2022 (#338)
yuzu-emu/yuzu-emu.github.io: site/content/entry/yuzu-progress-report-nov-2022/index.md

While most GPU drivers do indeed support these extensions, the NVIDIA drivers don't for Maxwell V1 and older cards. This means users trying to run `Pokémon Scarlet/Violet` on GPUs older than the 900 series wouldn’t get the game to display anything on screen, in either graphics API and any shader backend.
Must be part of the promised package that included Vulkan support for Fermi, pure smoke.

The way [byte[]](https://github.com/liamwhite) solved this limitation is by {{< gh-hovercard "9363" "translating the gl_Layer assignments with geometry shaders," >}} basically a brute-forced hardware accelerated way to replicate the missing extensions on older hardware.
This results in proper rendering for the decade-old Geforce Fermi, Kepler, and Maxwell V1 series GPUs, in either OpenGL and Vulkan.
Now an extra 6% of users can play the game.

{{< imgs
    "./710.png| The test subject, anyone can cook"
  >}}

&nbsp;

{{< imgs
    "./710fix.png| Old but gold! (Pokémon Scarlet)"
  >}}

Let’s switch for a second to CPU related issues.

> There were reports that certain moves like `tail whip` could cause the opposing Pokémon to outright vanish, or that changing directions very quickly could cause a soft-lock. 

Reviewer: ```suggestion
There were reports that certain moves like `Tail Whip` could cause the opposing Pokémon to outright vanish, or that changing directions very quickly could cause a soft-lock. 
```

{{< single-title-imgs-compare
    "Wasn't this game based in Spain? (Pokémon Scarlet)"
    "./amdbug.png"
    "./amdfix.png"
>}}

{{< gh-hovercard "9260" "By adding the missing flat decoration," >}} byte[] allows Radeon users to view the proper beautiful blue skies, in both Windows and Linux.

Ping-ponging back to CPU bugs:
Even with all the work so far, `Pokémon Scarlet/Violet` could periodically crash.
This gave us a sense of dread we haven’t felt since the release of `Pokémon Sword/Shield`.
Thankfully, our fear was unfounded, as only Windows users reported this issue. 
Linux users, for example Steam Deck users, didn’t crash at all.
This gave us the needed hook to reel the line of the cause, which ended up being a problem in dynarmic, {{< gh-hovercard "9271" "a stack misalignment in one of the memory accessors." >}}
A quick call to [Merry](https://github.com/merryhime) over the red phone, and the crashing is no more.

Back to the GPU, and this one was really annoying.
Your writer had to do an all-nighter running tests with [bunnei](https://github.com/bunnei) to figure it out.

> NPCs and characters could encounter vertex explosions at random, and while using high GPU accuracy would mitigate the problem to a certain extent, it was still very common in cities, particularly in the academy.

Reviewer: ```suggestion
NPCs and characters could encounter vertex explosions at random, and while using `High` GPU accuracy would mitigate the problem to a certain extent, it was still very common in cities, particularly in the academy.
```

This way, the game remains playable in Vulkan for AMD on Windows.

{{< imgs
    "./xc3.png| Always has great title screen music (Xenoblade Chronicles 3)"
  >}}

Here’s an interesting find from one of our *old* testers.
If you run `The Legend of Zelda: Breath of the Wild` in its base 1.0.0 version, recent changes caused distant trees to have a black flicker.
This behaviour was not present in an updated game, but a glitch is a glitch, and it deserves attention.

{{< single-title-imgs
    "The old result (left) has been reduced to 10 FPS to avoid triggering sensitive readers (The Legend of Zelda: Breath of the Wild)"
    "./botwbug.mp4"
    "./botwfix.mp4"
    >}}

vonchenplus found out that the order of the drawing commands was wrong in this case, and implemented a {{< gh-hovercard "9288" "more conservative drawing trigger mechanism." >}}
No other game is known to be affected by this change, but if there is, it’s fixed!

Your writer has been trying to slowly learn the ropes and decided to take on some small projects.
Too bad he’s also an idiot…

> The result of too many hours of suffering is the addition of a {{< gh-hovercard "9276" "sharpening slider for the FSR filter." >}} Also known by its long name, AMD FidelityFX Super Resolution 1.0.2.

Reviewer: ```suggestion
The result of too many hours of suffering is the addition of a {{< gh-hovercard "9276" "sharpening slider for the FSR filter." >}} Also known by its long name, `AMD FidelityFX Super Resolution 1.0.2`. 
```

> FSR between friends.

Reviewer: ```suggestion
However, it prefers being called FSR between friends.
```

{{< single-title-imgs
    "This is what's possible right now (Mario Kart 8 Deluxe & Super Mario Galaxy)"
    "./asahi.png"
    "./asahi2.png"
    >}}

This initial implementation adds support for 32-bit games only. 
As for 64-bit games (the most common), [an update needs to be merged to Dynarmic](https://github.com/merryhime/dynarmic/pull/719). 
Once it’s out, there will be no restrictions on which games can be executed.

{{< single-title-imgs
    "This is what will be possible in the near future (Super Mario 3D World + Bowser's Fury, Super Mario 3D All-Stars, & Super Mario Odyssey)"
    "./asahi64.png"
    "./asahi642.png"
    "./asahi643.png"
    >}}

Further improvements by byte[] includes {{< gh-hovercard "9215" "corrections on atomic store ordering" >}} to improve stability, {{< gh-hovercard "9234" "implement data cache management operations," >}} and the most fun one, {{< gh-hovercard "9289" "fix compilation for Apple Clang," >}} allowing yuzu to run directly on macOS!

{{< imgs
    "./macos.png| If only Apple wasn't a reptilian company and offered Vulkan support"
  >}}

> We have a very long road ahead to get full macOS support, but a journey of a thousand miles begins with a single step, says the old man.

Reviewer: ```suggestion
We have a very long road ahead to fully support macOS, but a journey of a thousand miles begins with a single step, says the old man.
```

This first implementation will only show up when a game requests it, a manual UI option will be added in the future.

{{< imgs
    "./cabinet.png| Images optional for now (Shovel Knight)"
  >}}

On the audio side of things, [Maide](https://github.com/Kelebek1) has been working on solving out-of-bounds crashing issues related to how the emulator and the Switch handle audio buffers.
Here’s the thing, the game decides how many buffers it wants to use, and has the freedom to use the buffers as it pleases.
As Maide says: “Put voice A in buffer 1 with voice B in buffer 2, mix them together into buffer 3. Put voice C in buffer 4 and mix buffer 3 and 4 into buffer 5, etc.”
At the end of the line, if you need to output to 2 channels, only two final buffers will be used, for 6 channels, 6 buffers, but there is no order to those selected buffers, no rule says that they must be buffer 0 and 1.
yuzu operated under the assumption that games would use buffers 0 to 5 for output, and that has been the case for the majority of games, but as always, there are exceptions, one being `The Legend of Zelda: Skyward Sword HD`.

> This particular game uses something similar to 8/9/12/13/10/11 for its 6 buffer outputs, or in proper hexadecilam values, 8/9/C/D/A/B.

Reviewer: ```suggestion
This particular game uses something similar to 8/9/12/13/10/11 for its 6 buffer outputs, or in proper hexadecimal values, 8/9/C/D/A/B.
```

The previous implementation took the first buffer as a starting point, buffer 0, and added +1 to fill the remaining 5 slots.

> That’s definitely not valid for Skyward Sword, the moment you want to access a buffer that is not part of those 6 first consecutive buffers, you get a crash for getting out-of-bounds, and this happens even with the first actual output buffer, at 8!

Reviewer: ```suggestion
That’s definitely not valid for `Skyward Sword`, the moment you want to access a buffer that is not part of those 6 first consecutive buffers, you get a crash for getting out-of-bounds, and this happens even with the first actual output buffer, at 8!
```

Fixing this behaviour by {{< gh-hovercard "9297" "checking the highest buffer and adding +1" >}} stops the game from crashing, and should also have the benefit of not outputting weird sounds, as the correct buffers will be selected.

As most of you know, our current file system emulation is far from ideal. 
Morph’s `Project Gaia` should address most issues we have with it once his rewrite is finished.
In the meantime, there’s nothing against fixing existing problems with it, and [v1993](https://github.com/v1993) provides a huge hand by addressing a case that could be invisible to the user, and really annoying as a result.
The emulator can auto-generate some of the keys required to decrypt and run Switch games.
The problem is that those auto-generated keys are sometimes worse than those the user provided during the dumping process, yet they still took precedence.
v1993 {{< gh-hovercard "9324" "flips this behaviour," >}} solving cases of games refusing to boot due to this conflict.

## User interface improvements

Localization is an ongoing effort, and anyone can contribute [here](https://www.transifex.com/yuzu-emulator/yuzu).

> This time the community {{< gh-hovercard "9166" "added support for another language, Ukrainian!" >}}

Reviewer: ```suggestion
This time the community added support for another language, {{< gh-hovercard "9166" "Ukrainian!" >}}
```

This addition is possible thanks to the work done  by [Docteh](https://github.com/Docteh) and GillianMC.
Keep at it guys, we're all with you.

Docteh also {{< gh-hovercard "9180" "fixed the translation of pop-up warnings" >}} that show up when trying to remove game content.

{{< single-title-imgs
    "Como corresponde"
    "./removebug.png"
    "./removefix.png"
>}}

No one wants to lose their saves, especially not if it's the user’s own fault.
As per the request of one such user, toastUnlimited added {{< gh-hovercard "9247" "a warning pop-up when a user attempts to delete a user profile." >}}
Because, yes, deleting a profile includes deleting its saves.

{{< imgs
    "./delete.png| The UI must also protect the users from themselves"
  >}}

> epicboy surprised everyone by slamming down the front door, throwing {{< gh-hovercard "9273" "per-game input profile support" >}} in our faces, and leaving, refusing to elaborate further.
> A real Chad.

Reviewer: ```suggestion
epicboy surprised everyone by slamming the front door open, throwing {{< gh-hovercard "9273" "per-game input profile support" >}} in our faces, and leaving, refusing to elaborate further.
A real Chad.
```

V1993 {{< gh-hovercard "9324" "flips this behaviour," >}} solving cases of games refusing to boot due to this conflict.

## User interface improvements

Localization is an ongoing effort, and anyone can contribute [here](https://www.transifex.com/yuzu-emulator/yuzu).
This time the community {{< gh-hovercard "9166" "added support for a new language, Ukrainian!" >}}
This addition is possible thanks to the work done  by [Docteh](https://github.com/Docteh) and GillianMC.
Keep at it, guys, we're all with you.

Docteh also {{< gh-hovercard "9180" "fixed the translation of pop-up warnings" >}} that show up when trying to remove game content.

{{< single-title-imgs
    "Como corresponde"
    "./removebug.png"
    "./removefix.png"
>}}

No one wants to lose their saves, especially not if it is the own user’s fault.
As per the request of an user, toastUnlimited added {{< gh-hovercard "9247" "a warning pop-up when an user attempts to delete an user profile." >}}
Because, yes, deleting a profile includes deleting its saves.

{{< imgs
    "./delete.png| The UI must also protect the users from themselves"
  >}}

> epicboy surprised everyone by slamming down the front door, throwing {{< gh-hovercard "9273" "per-game input profile support" >}} into our faces, and leaving,  refusing to elaborate further.
> A real Chad.

Reviewer: ```suggestion
epicboy surprised everyone by slamming down the front door, throwing {{< gh-hovercard "9273" "per-game input profile support" >}} in our faces, and leaving, refusing to elaborate further.
A real Chad.
```

{{< imgs
    "./input.png| Right-click game > Properties > Input Profiles"
  >}}
  
This feature uses previously created input profiles, make sure to create some in `Emulation > Configure... > Controls` beforehand.

## Future projects

> There’s a *lot* going on behind doors, but it’s too early to start teasing you.
> We can confirm that both `Project Gaia` and `Project Y.F.C. Part 2` are progressing healthily, but those are not the only very important changes in the oven.

Reviewer: ```suggestion
There’s a *lot* going on behind closed doors, but it’s too early to start teasing you.
We can confirm that both `Project Gaia` and `Project Y.F.C. Part 2` are progressing healthily, but those are not the only important changes in the oven.
```
